src/EEAlgorithmValencia.jl

```diff
@@ -0,0 +1,408 @@
+
+Precomputed Valencia cross-update variant: updates neighbour data for slot
+`i` and any affected neighbours using `E2p_scaled` and `beam_term`.
+"""
+@inline function update_nn_cross_arrays_precomputed!(nndist::AbstractVector,
+                                                     nni::AbstractVector,
+                                                     nx::AbstractVector, ny::AbstractVector,
+                                                     nz::AbstractVector,
+                                                     E2p_scaled::AbstractVector,
+                                                     beam_term::AbstractVector,
+                                                     dijdist::AbstractVector,
+                                                     i::Integer, N::Integer,
+                                                     dij_factor, β = 1.0, γ = 1.0, R = 4.0)
+    nndist_i = Inf
+    nni_i = i
+    @inbounds for j in 1:(i - 1)
+        this_metric = valencia_distance_inv_scaled_arrays(E2p_scaled, nx, ny, nz, i, j)
+        if this_metric < nndist_i
+            nndist_i = this_metric
+            nni_i = j
+        end
+        if this_metric < nndist[j]
+            nndist[j] = this_metric
+            nni[j] = i
+            dijdist[j] = this_metric
+            if beam_term[j] < dijdist[j]
+                dijdist[j] = beam_term[j]
+                nni[j] = 0
+            end
+        end
+    end
+    @inbounds for j in (i + 1):N
+        this_metric = valencia_distance_inv_scaled_arrays(E2p_scaled, nx, ny, nz, i, j)
+        if this_metric < nndist_i
+            nndist_i = this_metric
+            nni_i = j
+        end
+        if this_metric < nndist[j]
+            nndist[j] = this_metric
+            nni[j] = i
+            dijdist[j] = this_metric
+            if beam_term[j] < dijdist[j]
+                dijdist[j] = beam_term[j]
+                nni[j] = 0
+            end
+        end
+    end
+    # Finalize slot i
+    dijdist_i = valencia_distance_inv_scaled_arrays(E2p_scaled, nx, ny, nz, i, nni_i)
+    if beam_term[i] < dijdist_i
+        dijdist_i = beam_term[i]
+        nni_i = 0
+    end
+    nndist[i] = nndist_i
+    nni[i] = nni_i
+    dijdist[i] = dijdist_i
+end
+
+"""
+    _ee_genkt_algorithm_valencia(; particles::AbstractVector{EEJet}, algorithm::JetAlgorithm.Algorithm, p::Real, R=4.0, recombine=addjets, γ::Real=1.0, beta::Union{Real, Nothing}=nothing)
+
+Valencia-specialised implementation of the e+e- gen-kT clustering algorithm.
+This implementation precomputes scaled energy and beam-term arrays to speed up
+nearest-neighbour computations for the Valencia metric.
+"""
+function _ee_genkt_algorithm_valencia(; particles::AbstractVector{EEJet},
+                                      algorithm::JetAlgorithm.Algorithm, p::Real, R = 4.0,
+                                      recombine = addjets, γ::Real = 1.0,
+                                      beta::Union{Real, Nothing} = nothing)
```

Comment:
```suggestion
"""
    _ee_genkt_algorithm_valencia!(particles::AbstractVector{EEJet}; algorithm::JetAlgorithm.Algorithm, p::Real, R=4.0, recombine=addjets, γ::Real=1.0, beta::Union{Real, Nothing}=nothing)

Valencia-specialised implementation of the e+e- gen-kT clustering algorithm.
This implementation precomputes scaled energy and beam-term arrays to speed up
nearest-neighbour computations for the Valencia metric.
"""
function _ee_genkt_algorithm_valencia!(particles::AbstractVector{EEJet};
                                      algorithm::JetAlgorithm.Algorithm, p::Real, R = 4.0,
                                      recombine = addjets, γ::Real = 1.0,
                                      beta::Union{Real, Nothing} = nothing)
```

In #195 we renamed internal algorithms to consistently have `!` in name (since they are mutating) and take particles as positional argument